In [126]:
# !pip install numpy
# !pip install pandas
# !pip install matplotlib

In [127]:
import numpy as np;
import pandas as pd;
import matplotlib.pyplot as plt

In [13]:
award_info_df = pd.read_csv(
    './DataSets/awards_moms.csv')

registry_df = pd.read_csv(
    './DataSets/registry_game.csv')

mission_player_df = pd.read_csv(
    './DataSets/missions_players.csv')

In [21]:
award_info_cols = ["data_richness_prize", "id_player", "prize_name", "prize_points", "prize_type", "prize_format", "delivery_mode"]
registry_cols = ["id_player", "user_registration_date", "child_date_of_birth", "current_child_age", "child_age_registration_date", "province", "province_code_abbreviation", "municipality", "region"]
mission_player_cols = ["id_player",	"missionDetailId", "missionSchemaId", "type", "subType", "points", "created_at"]

In [22]:
award_info_df.columns = award_info_cols
registry_df.columns = registry_cols
mission_player_df.columns = mission_player_cols

In [23]:
award_info_df.head()

data_richness_prize  id_player                 prize_name  prize_points  \
0          2022-01-20     1672.0  Bicicletta Personalizzata         21000   
1          2021-10-20   537088.0  Velocipede Personalizzato         18000   
2          2021-10-08   537088.0  Velocipede Personalizzato         18000   
3          2022-01-27  1461448.0       Buono Sconto Photosì          3000   
4          2021-12-28  1633286.0     Buono Sconto Leolandia             0   

  prize_type prize_format delivery_mode  
0      basic     physical          home  
1      basic     physical          home  
2      basic     physical          home  
3       gift      digital         email  
4    special      digital         email

In [24]:
registry_df.head()

id_player user_registration_date child_date_of_birth  current_child_age  \
0    1878238             2022-06-16          2021-11-22               10.0   
1    1289508             2021-06-09          2021-03-03               18.0   
2    1170534             2018-10-19          2021-01-14               20.0   
3     964802             2021-03-02          2021-02-13               19.0   
4     549416             2020-03-24          2020-11-08               22.0   

   child_age_registration_date province province_code_abbreviation  \
0                          6.0  TARANTO                         TA   
1                          3.0    LUCCA                         LU   
2                        -27.0  MANTOVA                         MN   
3                          0.0  RAVENNA                         RA   
4                         -8.0    CUNEO                         CN   

   municipality          region  
0           NaN          PUGLIA  
1           NaN         TOSCANA  
2  RONCOFERRARO       LOMBARDIA  
3       RAVENNA  EMILIA-ROMAGNA  
4           BRA        PIEMONTE

In [25]:
mission_player_df.head()

id_player  missionDetailId  missionSchemaId     type subType  points  \
0    1105190     1.094383e+08            194.0  special     cib   500.0   
1    1895304     2.238056e+09            215.0  special     cib   100.0   
2    1879684     1.767847e+09            215.0  special     cib   100.0   
3    1874660     1.643138e+09            215.0  special     cib   100.0   
4    1521418     1.104016e+08            195.0  special     cib   900.0   

            created_at  
0  2021-10-06 16:21:26  
1  2022-07-02 18:37:15  
2  2022-06-17 15:03:46  
3  2022-06-13 11:42:48  
4  2021-10-26 16:36:30

In [177]:
print(mission_player_df[mission_player_df["missionSchemaId"] == 215]["points"].count())

10076


#### Clean data: replace points on the basis of majority by checking missionSchemaId

100 10076
2 33

In [27]:
def empty_cell(dataframe, column:str):
    count_nan_column = dataframe[column].isna().sum()
    print(f"Empty Rows Count of {column}: {count_nan_column}")

    dataframe = dataframe.dropna(subset=[column])
    return dataframe

In [120]:
def adjust_data(dataframe, source_column:str, target_column:str):
    print(source_column, target_column)
    # unique_source_column_data = dataframe[source_column].unique()
    demo = dataframe[dataframe[source_column] == "Buono Sconto Leolandia"]
    print(demo[target_column])
    print(demo[demo[target_column] == 0]["prize_type"].unique())
    print(demo[demo["prize_type"] == "special"][target_column].unique())
    # print(bicicletta_personalizzata)
    # demo = bicicletta_personalizzata.groupby(target_column).count()
    # print(demo)
    # print(bicicletta_personalizzata["prize_type"].nunique())
    # bicicletta_personalizzata_price = bicicletta_personalizzata[target_column].unique()
    # print(bicicletta_personalizzata_price)

    # return None
    # return dataframe

In [121]:
adjust_data(award_info_df, "prize_name", "prize_points")

prize_name prize_points
['special']
[0]


In [131]:
print(award_info_df.groupby(["prize_type", "prize_format", "delivery_mode"])["id_player"].count())

prize_type  prize_format  delivery_mode
basic       digital       email             1604
            physical      home             19380
gift        digital       email             8136
special     digital       email             1853
            physical      home              1367
Name: id_player, dtype: int64


In [135]:
print(award_info_df["prize_name"].unique())

['Bicicletta Personalizzata' 'Velocipede Personalizzato'
 'Buono Sconto Photosì' 'Buono Sconto Leolandia'
 'Buono sconto My Sweetie Pie' 'Primi Passi' 'Inventa favole Montessori'
 'Buono Sconto Clementoni' 'Calendario personalizzato'
 '10€ Buono sconto Pampers' 'Clinica Veterinaria' 'Cucina'
 'Registratore di cassa' 'Banco lavoro' 'Buono sconto Playmobil'
 '15€ Buono sconto Pampers' 'Casetta Pretty'
 'Tazza personalizzata con foto' 'Buono Sconto Sviluppofoto' 'Trattore'
 'Buono sconto Carrera' 'Casetta Pretty Personalizzata'
 'Buono Sconto Zac4Kids' 'Buono Sconto Tonki' 'Buono Sconto Landoloux'
 '20€ Buono sconto Pampers' 'Cavallino gonfiabile'
 '5€ Buono sconto Pampers' 'Bicicletta' 'Pizzeria' 'Supermercato'
 'Ciondolino Manufatto' 'Ruspa' 'Doctor on call' 'Valigetta del dottore'
 'A pesca' 'Flashcards della buonanotte' 'Buono sconto Semiperdo'
 'Buono sconto MyMi' 'Quad Racer Elettrico' 'Kit Green Generation'
 'Playset Aereo' '10€ Coupon Hero*' 'Velocipede' 'Pochette summer'
 'Primi 

Customer get a reward

In [157]:
print(award_info_df[(award_info_df["prize_name"].str.contains("Buono")) & (award_info_df["prize_name"].str.contains("Pampers"))].groupby("prize_type").count())

            data_richness_prize  id_player  prize_name  prize_points  \
prize_type                                                             
basic                      2163       2163        2163          2163   

            prize_format  delivery_mode  
prize_type                               
basic               2163           2163  


### Scenario: to redeem the prize, a user should have more points (for eg: for pamper minimum user have spent is 5500 points where as the maximum is 14500)

In [165]:
award_info_df[(award_info_df["prize_name"].str.contains("Buono")) & (award_info_df["prize_name"].str.contains("Pampers"))].groupby(["id_player", "prize_type"]).sum()

C:\Users\zaime\AppData\Local\Temp\ipykernel_13864\2239440850.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  award_info_df[(award_info_df["prize_name"].str.contains("Buono")) & (award_info_df["prize_name"].str.contains("Pampers"))].groupby(["id_player", "prize_type"]).sum()


prize_points
id_player prize_type              
1348.0    basic              14500
2002.0    basic               5500
2956.0    basic              29000
3072.0    basic               9000
3872.0    basic               9000
...                            ...
1652278.0 basic               5500
1654324.0 basic               5500
1666806.0 basic               5500
1671910.0 basic              29000
1683588.0 basic               9000

[1952 rows x 1 columns]

In [169]:
award_info_df[(award_info_df["prize_name"].str.contains("Buono")) & (award_info_df["prize_name"].str.contains("Pampers"))].groupby(["id_player", "prize_type"]).count()

data_richness_prize  prize_name  prize_points  \
id_player prize_type                                                  
1348.0    basic                         1           1             1   
2002.0    basic                         1           1             1   
2956.0    basic                         2           2             2   
3072.0    basic                         1           1             1   
3872.0    basic                         1           1             1   
...                                   ...         ...           ...   
1652278.0 basic                         1           1             1   
1654324.0 basic                         1           1             1   
1666806.0 basic                         1           1             1   
1671910.0 basic                         2           2             2   
1683588.0 basic                         1           1             1   

                      prize_format  delivery_mode  
id_player prize_type                               
1348.0    basic                  1              1  
2002.0    basic                  1              1  
2956.0    basic                  2              2  
3072.0    basic                  1              1  
3872.0    basic                  1              1  
...                            ...            ...  
1652278.0 basic                  1              1  
1654324.0 basic                  1              1  
1666806.0 basic                  1              1  
1671910.0 basic                  2              2  
1683588.0 basic                  1              1  

[1952 rows x 5 columns]

### Scenario: some users have redeemed only special prize type where the prize point is 0

In [168]:
award_info_df.groupby(["id_player", "prize_type"]).sum()

C:\Users\zaime\AppData\Local\Temp\ipykernel_13864\3005341398.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  award_info_df.groupby(["id_player", "prize_type"]).sum()


prize_points
id_player prize_type              
1190.0    basic               6000
1348.0    basic              14500
1380.0    basic              12000
          gift                9000
          special                0
...                            ...
1930226.0 special                0
1930896.0 special                0
1931046.0 basic              14500
1933846.0 special                0
1953116.0 basic               7000

[24490 rows x 1 columns]

In [170]:
award_info_df.groupby(["id_player", "prize_type"]).count()

data_richness_prize  prize_name  prize_points  \
id_player prize_type                                                  
1190.0    basic                         1           1             1   
1348.0    basic                         1           1             1   
1380.0    basic                         2           2             2   
          gift                          2           2             2   
          special                       1           1             1   
...                                   ...         ...           ...   
1930226.0 special                       1           1             1   
1930896.0 special                       1           1             1   
1931046.0 basic                         1           1             1   
1933846.0 special                       1           1             1   
1953116.0 basic                         1           1             1   

                      prize_format  delivery_mode  
id_player prize_type                               
1190.0    basic                  1              1  
1348.0    basic                  1              1  
1380.0    basic                  2              2  
          gift                   2              2  
          special                1              1  
...                            ...            ...  
1930226.0 special                1              1  
1930896.0 special                1              1  
1931046.0 basic                  1              1  
1933846.0 special                1              1  
1953116.0 basic                  1              1  

[24490 rows x 5 columns]

In [ ]:
['Customized Bicycle', 
'Customized Velocipede',
'Photosì Discount Voucher', 
'Leolandia Discount Voucher',
'My Sweetie Pie Discount Voucher', 
'First Steps', 
'Montessori Fairy Tales Invent',
'Clementoni Discount Voucher', 
'Customized Calendar',
'10€ Pampers Discount Voucher', 
'Veterinary Clinic', 
'Kitchen',
'Cash Register', 
'Workbench', 
'Playmobil Discount Voucher',
'15€ Pampers Discount Voucher', 
'Pretty House',
'Personalized Photo Mug', 
'Sviluppofoto Discount Voucher', 
'Tractor',
'Carrera Discount Voucher', 
'Personalized Pretty House',
'Zac4Kids Discount Voucher', 
'Tonki Discount Voucher', 
'Landoloux Discount Voucher',
'20€ Pampers Discount Voucher', 
'Inflatable Pony',
'5€ Pampers Discount Voucher', 
'Bicycle', 
'Pizzeria', 
'Supermarket',
'Handmade Ciondolino', 
'Bulldozer', 
'Doctor on call', 
'Doctors Briefcase',
'Fishing', 
'Bedtime Flashcards', 
'Semiperdo Discount Voucher',
'MyMi Discount Voucher', 
'Electric Quad Racer', 
'Kit Green Generation',
'Airplane Playset', 
'10€ Hero Coupon*', 
'Velocipede', 
'Summer Pouch',
'Personalized First Steps', 
'1 Month of FitPrime TV',
"Mother's Day: Pouch", 
'Desk', 
'Xylophone',
'Abrick 100 Bricks', 
'Screwdriver Workbench', 
'School', 
'Scooter',
'HERO Solo Cookbook', 
'Ukulele', 
'Portable Puppy',
'Tantosvago Discount Voucher', 
'Number Flashcards', 
'Coffee Machine',
'10€ Pampers Coupon - Ticket Coupon', 
'Donchisciotte Discount Voucher',
'Duck Playset', 
'Kitchen Briefcase', 
'3 Months of FitPrime TV',
'1 Month of Online Gym', 
'My Montessori House Games', 
'Market',
'Tool Belt', 
'Block & Roll', 
'Lion House']

In [ ]:
['Bicicletta Personalizzata', 'Velocipede Personalizzato',
 'Buono Sconto Photosì', 'Buono Sconto Leolandia',
 'Buono sconto My Sweetie Pie', 'Primi Passi', 'Inventa favole Montessori',
 'Buono Sconto Clementoni', 'Calendario personalizzato',
 '10€ Buono sconto Pampers', 'Clinica Veterinaria', 'Cucina',
 'Registratore di cassa', 'Banco lavoro', 'Buono sconto Playmobil',
 '15€ Buono sconto Pampers', 'Casetta Pretty',
 'Tazza personalizzata con foto', 'Buono Sconto Sviluppofoto', 'Trattore',
 'Buono sconto Carrera', 'Casetta Pretty Personalizzata',
 'Buono Sconto Zac4Kids', 'Buono Sconto Tonki', 'Buono Sconto Landoloux',
 '20€ Buono sconto Pampers', 'Cavallino gonfiabile',
 '5€ Buono sconto Pampers', 'Bicicletta', 'Pizzeria', 'Supermercato',
 'Ciondolino Manufatto', 'Ruspa', 'Doctor on call', 'Valigetta del dottore',
 'A pesca', 'Flashcards della buonanotte', 'Buono sconto Semiperdo',
 'Buono sconto MyMi', 'Quad Racer Elettrico', 'Kit Green Generation',
 'Playset Aereo', '10€ Coupon Hero*', 'Velocipede', 'Pochette summer',
 'Primi Passi Personalizzata', '1 mese di FitPrime TV',
 'Festa della Mamma: Pochette', 'Scrivania', 'Xilofono',
 'Abrick 100 mattoncini', 'Banchetto Svita Avvita', 'Scuola', 'Monopattino',
 'Ricettario HERO Solo', 'Ukulele', 'Cagnolino Trasportabile',
 'Buono Sconto Tantosvago', 'Flashcards number', 'Macchinetta del Caffe',
 '10€ Coupon Pampers - Ticket Coupon', 'Buono sconto Donchisciotte',
 'Playset Paperelle', 'Valigetta cucina', '3 mesi di FitPrime TV',
 'Un mese di palestra online', 'La Mia casa Giochi Montessori', 'Mercatino',
 'Cintura Attrezzi', 'Block & roll', 'Casetta Leone']

In [ ]:
def feature_engineering(registry_game):

    # Feature engineering
    # create age groups
    bins = [0, 18, 30, 40, 50, 60, 100]
    labels = ['0-18', '19-30', '31-40', '41-50', '51-60', '61+']
    registry_game['age_group'] = pd.cut(
        registry_game['ETA_MM_BambinoTODAY'], bins=bins, labels=labels)

    registry_game_result = registry_game.groupby('age_group')[
        'id_player'].count()
    print(registry_game_result)
    registry_game_result.plot(kind='bar')
    plt.xlabel('TIER')
    plt.ylabel('Customers')
    plt.title('Count of Y grouped by X')
    plt.show()

    # view the data
    print(registry_game_result.head())

In [28]:
award_info_df = empty_cell(award_info_df, "id_player")
award_info_df.head()

Empty Rows Count of id_player: 0


data_richness_prize  id_player                 prize_name  prize_points  \
0          2022-01-20     1672.0  Bicicletta Personalizzata         21000   
1          2021-10-20   537088.0  Velocipede Personalizzato         18000   
2          2021-10-08   537088.0  Velocipede Personalizzato         18000   
3          2022-01-27  1461448.0       Buono Sconto Photosì          3000   
4          2021-12-28  1633286.0     Buono Sconto Leolandia             0   

  prize_type prize_format delivery_mode  
0      basic     physical          home  
1      basic     physical          home  
2      basic     physical          home  
3       gift      digital         email  
4    special      digital         email

year exceeding 2022 exclude the data

In [29]:
registry_df = empty_cell(registry_df, "id_player")
registry_df.head()

Empty Rows Count of id_player: 0


id_player user_registration_date child_date_of_birth  current_child_age  \
0    1878238             2022-06-16          2021-11-22               10.0   
1    1289508             2021-06-09          2021-03-03               18.0   
2    1170534             2018-10-19          2021-01-14               20.0   
3     964802             2021-03-02          2021-02-13               19.0   
4     549416             2020-03-24          2020-11-08               22.0   

   child_age_registration_date province province_code_abbreviation  \
0                          6.0  TARANTO                         TA   
1                          3.0    LUCCA                         LU   
2                        -27.0  MANTOVA                         MN   
3                          0.0  RAVENNA                         RA   
4                         -8.0    CUNEO                         CN   

   municipality          region  
0           NaN          PUGLIA  
1           NaN         TOSCANA  
2  RONCOFERRARO       LOMBARDIA  
3       RAVENNA  EMILIA-ROMAGNA  
4           BRA        PIEMONTE

2020 - 11 = 22
2020 - 12 = 21
2022 - 01 = 9
2022 - 10 - 01

Determine:
Current Month = Sep - Oct
Current Year = 2022

Date cannot determine

In [172]:
registry_df.groupby(["child_date_of_birth", "current_child_age"]).count()

id_player  user_registration_date  \
child_date_of_birth current_child_age                                      
2019-09-13           36.0                     21                      21   
2019-09-14           36.0                     25                      25   
2019-09-15           36.0                     35                      35   
2019-09-16           36.0                     28                      28   
2019-09-17           36.0                     30                      30   
...                                          ...                     ...   
2023-05-27          -8.0                       1                       1   
2023-05-28          -8.0                       1                       1   
2023-06-02          -9.0                       1                       1   
2023-06-04          -9.0                       1                       1   
2023-06-12          -9.0                       1                       1   

                                       child_age_registration_date  province  \
child_date_of_birth current_child_age                                          
2019-09-13           36.0                                       21        20   
2019-09-14           36.0                                       25        23   
2019-09-15           36.0                                       35        34   
2019-09-16           36.0                                       28        27   
2019-09-17           36.0                                       30        30   
...                                                            ...       ...   
2023-05-27          -8.0                                         1         1   
2023-05-28          -8.0                                         1         1   
2023-06-02          -9.0                                         1         1   
2023-06-04          -9.0                                         1         1   
2023-06-12          -9.0                                         0         1   

                                       province_code_abbreviation  \
child_date_of_birth current_child_age                               
2019-09-13           36.0                                      19   
2019-09-14           36.0                                      20   
2019-09-15           36.0                                      33   
2019-09-16           36.0                                      24   
2019-09-17           36.0                                      28   
...                                                           ...   
2023-05-27          -8.0                                        1   
2023-05-28          -8.0                                        1   
2023-06-02          -9.0                                        1   
2023-06-04          -9.0                                        0   
2023-06-12          -9.0                                        1   

                                       municipality  region  
child_date_of_birth current_child_age                        
2019-09-13           36.0                        17      20  
2019-09-14           36.0                        20      23  
2019-09-15           36.0                        28      34  
2019-09-16           36.0                        21      27  
2019-09-17           36.0                        25      30  
...                                             ...     ...  
2023-05-27          -8.0                          0       1  
2023-05-28          -8.0                          1       1  
2023-06-02          -9.0                          1       1  
2023-06-04          -9.0                          1       1  
2023-06-12          -9.0                          1       1  

[1336 rows x 7 columns]

## child dob is exceeding today's date

2023-06-12	-9.0	1	1	0	1	1	1	1

id_player	user_registration_date	child_date_of_birth	current_child_age	child_age_registration_date
1170534	       2018-10-19	              2021-01-14	    20.0	               -27.0                           Invalid
549416	        2020-03-24	              2020-11-08	    22.0	                 -8.0                           Valid

If exceeding 9 months exclude the data
else include it

#### User region eda

In [30]:
mission_player_df = empty_cell(mission_player_df, "id_player")
mission_player_df.head()

Empty Rows Count of id_player: 0


id_player  missionDetailId  missionSchemaId     type subType  points  \
0    1105190     1.094383e+08            194.0  special     cib   500.0   
1    1895304     2.238056e+09            215.0  special     cib   100.0   
2    1879684     1.767847e+09            215.0  special     cib   100.0   
3    1874660     1.643138e+09            215.0  special     cib   100.0   
4    1521418     1.104016e+08            195.0  special     cib   900.0   

            created_at  
0  2021-10-06 16:21:26  
1  2022-07-02 18:37:15  
2  2022-06-17 15:03:46  
3  2022-06-13 11:42:48  
4  2021-10-26 16:36:30

In [182]:
app_access_df = empty_cell(empty_cell(pd.read_csv("./DataSets/app_accesses.csv"), "id_player"), "source")

app_access_df.head()

Empty Rows Count of id_player: 0
Empty Rows Count of source: 0


id_player source           updated_at
0  1029996.0    app  2022-07-14 14:07:21
1  1704016.0    app  2022-07-14 14:07:54
2  1705608.0    app  2022-07-14 14:11:18
3   488940.0    app  2022-07-14 14:56:47
4  1747078.0    app  2022-07-14 15:00:07

In [181]:
print(app_access_df.groupby("source").count())

        id_player  updated_at
source                       
app        970177      970177
sito         5771        5771


##### updated at if exceeding current year (2022) and current month exclude it

In [183]:
product_ean_conv_df = pd.read_csv("./DataSets/product_ean_conversion.csv")

product_ean_conv_df.head()

EAN                REFERENZA_DES         SEGMENTO_DES OCCUSO_DES  \
0  4015400327776         Progressi Newborn CP            Progressi  Pannolini   
1  4015400471202            Progressi Mini CP            Progressi  Pannolini   
2  4015400534600              Progressi Micro            Progressi  Pannolini   
3  8001090900180  Progressi Mutandino Midi CP  Progressi Mutandino  Pannolini   
4  8001090900227  Progressi Mutandino Maxi CP  Progressi Mutandino  Pannolini   

    TIER  
0  TIER1  
1  TIER1  
2  TIER1  
3  TIER1  
4  TIER1

In [184]:
product_loaded_df = pd.read_csv("./DataSets/product_loaded.csv")

product_loaded_df.head()

id_player  missionDetail  points           created_at            EAN
0     687244            0.0     150  2021-02-02 11:41:17  8001480300774
1     488574            0.0     200  2021-02-02 11:51:37  8001480301955
2     269880            0.0    1200  2021-03-11 15:50:32  8001480301931
3     955870            0.0     200  2021-03-11 15:51:44  4015400327776
4     471766            0.0     150  2021-03-11 15:55:58  8001480092204

In [185]:
award_info_cols

['data_richness_prize',
 'id_player',
 'prize_name',
 'prize_points',
 'prize_type',
 'prize_format',
 'delivery_mode']